File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    
    idx = curr = 0
    def dfs(nums, curr, target, idx):
        
        if len(nums) == idx:
            return 1 if curr == target else 0
        
        return dfs(numbers, curr - nums[idx], target, idx + 1) \
          + dfs(numbers, curr + nums[idx], target, idx + 1)
    
    return dfs(numbers, curr, target, idx)

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    static int cnt = 0;
    public int solution(int[] numbers, int target) {
        dfs(0, 0, numbers, target);
        return cnt;
    }
    
    public static void dfs(int depth, int num, int[] numbers, int target) {
        
        if (depth == numbers.length) {
            if(num == target) {
                cnt++;
            }
            return;
        }
      
        dfs(depth + 1, num + numbers[depth], numbers, target);
        dfs(depth + 1, num - numbers[depth], numbers, target);

    }
}